### PR TITLE
set default values for ulimit

### LINKED ISSUE #5817: Make default ulimit unlimited for containerVM
Towards #5353

Update the base ulimit config in the bootstrap template to be unlimited. There's little utility in it in a single user system other than to force a soft failure of specific type on resource exhaustion.
This is an alteration to the _default_ configuration - existing limits can easily be applied in the same way unlimited could be configured currently.
This does close or address #5353 but makes the interim behaviour more friendly to normal use-cases.

Raised for 1.2 inclusion as the result of a triage pass on customer found issues, hence high.

### PR DESCRIPTION
**Pull request by jialin-li**
fixes #5817 

Changes default ulimit to
```
ulimit -a
-f: file size (blocks)             unlimited
-t: cpu time (seconds)             unlimited
-d: data seg size (kb)             unlimited
-s: stack size (kb)                unlimited
-c: core file size (blocks)        unlimited
-m: resident set size (kb)         unlimited
-l: locked memory (kb)             unlimited
-p: processes                      unlimited
-n: file descriptors               1048576
-v: address space (kb)             unlimited
-w: locks                          unlimited
-e: scheduling priority            0
-r: real-time priority             0
```

### CODE REVIEW

File: lib/tether/ops_linux.go [hunk 1]
```diff
@@ -73,6 +73,12 @@ const (
 	nfsFileSystemType  = "nfs"
 	ext4FileSystemType = "ext4"
 	bridgeTableNumber  = 201
+
+	defaultNOFILE = 1024 * 1024
+	defaultULimit = ^uint64(0)
+	// can replace with 'golang.org/x/sys/unix' package once vendor code is updated
```
Comment (matthewavery) [inline lib/tether/ops_linux.go]: We should merge your vendor update and then change this before merging this

File: lib/tether/ops_linux.go [hunk 2]
```diff
@@ -73,6 +73,12 @@ const (
 	nfsFileSystemType  = "nfs"
 	ext4FileSystemType = "ext4"
 	bridgeTableNumber  = 201
+
+	defaultNOFILE = 1024 * 1024
```
Comment (matthewavery) [inline lib/tether/ops_linux.go]: please add brief comments above all of these constant values to describe what they do and why they were picked :)

File: lib/tether/ops_linux.go (hunk 1 above)
Comment (hickeng) [inline lib/tether/ops_linux.go]: concur - let's get the vendor change merged and update this.

File: lib/tether/ops_linux.go [hunk 3]
```diff
@@ -929,6 +935,9 @@ func (t *BaseOperations) Setup(config Config) error {
 		return err
 	}
 
+	// NOTE: ulimit default values should change when we support ulimit configuration
+	setupDefaultULimit()
```
Comment (hickeng) [inline lib/tether/ops_linux.go]: This should be in `cmd/tether/ops_linux.go` not in the common `/lib/tether`. default ulimit's in the appliance will still be applied via `/etc/security/limits*`

File: lib/tether/ops_linux.go [hunk 4]
```diff
@@ -1125,3 +1134,36 @@ func handleUtilityExit(t *BaseOperations, pid, exitCode int) bool {
 
 	return true
 }
+	// set ulimit fields to unlimited as their default value
+func setupDefaultULimit() {
+	var rLimit syscall.Rlimit
+
+	rLimit.Max = defaultNOFILE
```
Comment (hickeng) [inline lib/tether/ops_linux.go]: comment that NOFILE doesn't support `defaultULimit` as a value

File: lib/tether/ops_linux.go [hunk 5]
```diff
@@ -929,6 +935,9 @@ func (t *BaseOperations) Setup(config Config) error {
 		return err
 	}
 
+	// NOTE: ulimit default values should change when we support ulimit configuration
```
Comment (matthewavery) [inline lib/tether/ops_linux.go]: is there a ticket open for this?

Comment (jialin-li) [inline lib/tether/ops_linux.go]: I believe it's #5353. This is the first step to address 5353 if we decide to support ulimit.

File: cmd/tether/ops_linux.go [hunk 6]
```diff
@@ -37,7 +38,13 @@ import (
 	"github.com/vmware/vic/pkg/trace"
 )
 
-const runMountPoint = "/run"
+const (
+	runMountPoint = "/run"
+
+	// default values to set for ulimit fields
+	defaultNOFILE = 1024 * 1024
+	defaultULimit = ^uint64(0)
```
Comment (caglar10ur) [inline cmd/tether/ops_linux.go]: this is nice but please use math.MaxUInt64 for the readers sake :)

File: cmd/tether/ops_linux.go [hunk 7]
```diff
@@ -126,9 +133,47 @@ func (t *operations) Setup(sink tether.Config) error {
 		}
 	}
 
+	// NOTE: ulimit default values should change when we support ulimit configuration
+	setupDefaultULimit()
```
Comment (caglar10ur) [inline cmd/tether/ops_linux.go]: maybe ApplyLimits or ApplyDefaultLimits?

File: cmd/tether/ops_linux.go [hunk 8]
```diff
@@ -126,9 +133,47 @@ func (t *operations) Setup(sink tether.Config) error {
...
 }
 
+// set ulimit fields to unlimited as their default value
+func setupDefaultULimit() {
+	var rLimit syscall.Rlimit
+
+	// NOFILE does not support defaultULimit as a value due to kernel restriction on number of open files
+	rLimit.Max = defaultNOFILE
+	rLimit.Cur = rLimit.Max
+	err := syscall.Setrlimit(syscall.RLIMIT_NOFILE, &rLimit)
```
Comment (caglar10ur) [inline cmd/tether/ops_linux.go]: just cosmetic but all those if statements can be collapsed into single line form (if err:= ...; err != nil {...})

File: cmd/tether/ops_linux.go (hunk 6 above)
Comment (jialin-li) [inline cmd/tether/ops_linux.go]: done!

File: cmd/tether/ops_linux.go (hunk 7 above)
Comment (jialin-li) [inline cmd/tether/ops_linux.go]: changed :)

File: cmd/tether/ops_linux.go (hunk 8 above)
Comment (jialin-li) [inline cmd/tether/ops_linux.go]: done!

**Pull request merged**